### PR TITLE
feat: std::format support for dpp::snowflake

### PR DESCRIPTION
**Pull request by ruslan-ilesik**
This pr adds std::format support for dpp::snowflake, if we do not have c++20 or have no include for <format> it will be automatically disabled for compilation.

Test case was implemented to check formatting. 

## Code change checklist

- [X] I have ensured that all methods and functions are **fully documented** using doxygen style comments.
- [X] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [X] I tested that my change works before raising the PR.
- [X] I have ensured that I did not break any existing API calls.
- [X] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

### CODE REVIEW

File: src/unittest/test.cpp [hunk 1]
```diff
@@ -379,6 +379,30 @@ Markdown lol \\|\\|spoiler\\|\\| \\~\\~strikethrough\\~\\~ \\`small \\*code\\* b
 		set_test(SNOWFLAKE, success);
 	}
 
+        { // test snowflake: std::format support
+
+                #if __cplusplus >= 202002L
```
Comment (braindigitalis) [inline src/unittest/test.cpp:384]: DRY,  I'm concerned someone may edit one of these checks and not the other, can we have it purely in one place please, maybe used to set our own define called DPP_HAS_FORMAT

**Review by braindigitalis — CHANGES_REQUESTED** [review]

File: include/dpp/snowflake.h [hunk 2]
```diff
@@ -281,3 +281,32 @@ struct std::hash<dpp::snowflake>
 		return std::hash<uint64_t>{}(s.value);
 	}
 };
+
+#if __cplusplus >= 202002L
+//fix for https://github.com/llvm/llvm-project/issues/77773 and
+// apple support on clang 14 https://developer.apple.com/xcode/cpp/
+#if (defined(__cpp_lib_format) || \
```
[the remark this answers is not in the document]
Comment (ruslan-ilesik) [reply include/dpp/snowflake.h:288]: Do we have runners with them to check?

[the remark this answers is not in the document]
Comment (braindigitalis) [reply include/dpp/snowflake.h:288]: of course but they don't run unit tests

File: src/unittest/test.cpp (hunk 1 above)
Comment (ruslan-ilesik) [reply src/unittest/test.cpp]: Done

File: include/dpp/snowflake.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ruslan-ilesik) [reply include/dpp/snowflake.h]: done

File: CMakeLists.txt [hunk 3]
```diff
@@ -115,3 +115,48 @@ else()
...
+
+	)
+	set(CMAKE_REQUIRED_FLAGS ${temp_flags})
+	if(HAS_FORMAT)
+		message("-- INFO: Found std::format support")
+		target_compile_definitions(dpp PUBLIC DPP_HAS_FORMAT)
+	else ()
+		message("-- INFO: std::format support not found")
+	endif()
+endif ()
```
Comment (Jaskowicz1) [inline CMakeLists.txt:162]: Add an empty line after this please!

File: include/dpp/snowflake.h [hunk 4]
```diff
@@ -26,6 +26,11 @@
 #include <cstdint>
 #include <type_traits>
 
+#ifdef DPP_HAS_FORMAT
+#include <format>
+#endif
+
+
```
Comment (Jaskowicz1) [inline include/dpp/snowflake.h:33]: Double empty line not needed

File: include/dpp/snowflake.h [hunk 5]
```diff
@@ -281,3 +286,23 @@ struct std::hash<dpp::snowflake>
...
+        constexpr typename TP::iterator parse(TP& ctx) {
+                return ctx.begin();
+        }
+
+        template<class TF>
+        typename TF::iterator format(const dpp::snowflake& snowflake, TF& ctx) const {
+                return std::format_to(ctx.out(), "{}", snowflake.str());
+        }
+};
+#endif //DPP_HAS_FORMAT
```
Comment (Jaskowicz1) [inline include/dpp/snowflake.h:308]: Empty line after this too.

**Review by Jaskowicz1 — CHANGES_REQUESTED** [review]

**Comment by Henonicks** [comment]
tested this and got
`snowflake: std::format support                               PASS`
at the end

**Comment by ruslan-ilesik** [comment]
So with last commit i have introduced usage of Cmake feature called [configure_file](https://cmake.org/cmake/help/v3.0/command/configure_file.html) which allows us autogenerate headers with some definitions when build dpp, this way we can remove need to define DPP_HAS_FORMAT as it will be defined in it when compiling it.

 In theory same way we can eliminate need to specify -DDPP_CORO when compiling bot with dpp enabled in lib.

File: CMakeLists.txt [hunk 6]
```diff
@@ -115,3 +115,53 @@ else()
...
+
+configure_file(
+		${CMAKE_CURRENT_SOURCE_DIR}/include/dpp/autoconfig.h.in
+		${CMAKE_CURRENT_SOURCE_DIR}/include/dpp/autoconfig.h
+		@ONLY
+)
```
Comment (Jaskowicz1) [inline CMakeLists.txt:167]: Add a blank line after this.

**Review by Jaskowicz1 — CHANGES_REQUESTED** [review]

**Review by braindigitalis — CHANGES_REQUESTED** [review]
theres a massive problem with this!
If cmake detects that C++20 is available it will always enable std::format. This means the user is **forced to use -std=c++20** for all their bots, even where they dont want to use C++20. This fundamentally breaks compatibility.
Nothing tells the user they must use C++20 anywhere, and furthermore, if they have a need for C++17 they simply cannot ever use C++17 any more.

**Comment by ruslan-ilesik** [comment]
If lib is build with c++20 then ye, bots will be required to use c++20, I can easily fix it by wrapping cmake define into check of c++ version.

**Review by Jaskowicz1 — APPROVED** [review]
I'm happy with it but need to see what @braindigitalis thinks about the `autoconfig.h` idea.

**Comment by ruslan-ilesik** [comment]
I have tested changes locally on my wsl2 (ubuntu) and tests passed, also compiling bot with -DPP_FORMATTERS and without had required behavior

File: library/CMakeLists.txt [hunk 7]
```diff
@@ -351,7 +355,7 @@ if (DPP_BUILD_TEST)
 			set (testsrc "")
 			file(GLOB testsrc "${modules_dir}/${testname}/*.cpp")
 			add_executable(${testname} ${testsrc})
-			if (DPP_CORO)
+			if (DPP_CORO OR DPP_FORMATTERS)
```
Comment (braindigitalis) [inline library/CMakeLists.txt:358]: these same changes need applying to library-vcpkg/CMakeLists.txt

**Review by braindigitalis — CHANGES_REQUESTED** [review]

**Review by braindigitalis — APPROVED** [review]

**Pull request merged**